homeassistant/components/matter/__init__.py

```diff
@@ -0,0 +1,234 @@
+from homeassistant.helpers import device_registry as dr
+from homeassistant.helpers.service import async_register_admin_service
+
+from .adapter import MatterAdapter, get_matter_store
+from .const import DOMAIN
+from .device_platform import DEVICE_PLATFORM
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Matter from a config entry."""
+    matter = Matter(MatterAdapter(hass, entry))
+    try:
+        await matter.connect()
+    except CannotConnect as err:
+        raise ConfigEntryNotReady("Failed to connect to matter server") from err
+
+    except Exception as err:
+        matter.adapter.logger.exception("Failed to connect to matter server")
+        raise ConfigEntryNotReady(
+            "Unknown error connecting to the Matter server"
+        ) from err
+
+    async def on_hass_stop(event: Event) -> None:
+        """Handle incoming stop event from Home Assistant."""
+        await matter.disconnect()
+
+    entry.async_on_unload(
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, on_hass_stop)
+    )
+
+    matter.listen()
+    try:
+        async with async_timeout.timeout(30):
+            await matter.driver_ready.wait()
+    except asyncio.TimeoutError as err:
+        raise ConfigEntryNotReady("Matter driver not ready") from err
+
+    if DOMAIN not in hass.data:
+        hass.data[DOMAIN] = {}
+        _async_init_services(hass)
+
+    hass.data[DOMAIN][entry.entry_id] = matter
+
+    await hass.config_entries.async_forward_entry_setups(entry, DEVICE_PLATFORM)
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, DEVICE_PLATFORM)
+
+    if unload_ok:
+        matter: Matter = hass.data[DOMAIN].pop(entry.entry_id)
+        await matter.disconnect()
+
+    return unload_ok
+
+
+async def async_remove_entry(hass: HomeAssistant, entry: ConfigEntry) -> None:
+    """Config entry is being removed."""
+    # Remove storage file.
+    storage_path = get_matter_store(hass, entry).path
+    await hass.async_add_executor_job(Path(storage_path).unlink)
+
+
+async def async_remove_config_entry_device(
+    hass: HomeAssistant, config_entry: ConfigEntry, device_entry: dr.DeviceEntry
+) -> bool:
+    """Remove a config entry from a device."""
+    unique_id = None
+
+    for ident in device_entry.identifiers:
+        if ident[0] == DOMAIN:
+            unique_id = ident[1]
+            break
+
+    if not unique_id:
+        return True
+
+    matter: Matter = hass.data[DOMAIN][config_entry.entry_id]
+
+    for node in matter.get_nodes():
+        if node.unique_id == unique_id:
+            await matter.delete_node(node.node_id)
+            break
+
+    return True
+
+
+@callback
+def _async_init_services(hass: HomeAssistant) -> None:
+    """Init services."""
+
+    async def commission(call: ServiceCall) -> None:
```

Comment:
I think we should not merge in the services but leave them out for this first PR.
We just need to make sure that we add the web socket commands.
Do we have something like a feature toggle to allow us to add Matter in smaller chunks (PR's) to core and at some point mark it as available ?

Reply:
I added the services following Paulus instructions in the clickup task.

Reply:
I'm using a separate target branch so that we can make multiple PRs before we merge to dev.